src/main/java/com/mraof/minestuck/item/loot/MSLootEvents.java

```diff
@@ -2,28 +2,137 @@
 	{
-		if(LOOT_INJECT.contains(event.getName()))
+		inject(event, BLANK_DISK_DUNGEON_LOOT_INJECT, MSLootTables.BLANK_DISK_DUNGEON_LOOT_INJECT, "blank_disk_dungeon_inject");
+		inject(event, SBURB_CODE_LIBRARY_LOOT_INJECT, MSLootTables.SBURB_CODE_LIBRARY_LOOT_INJECT, "sburb_code_library_inject");
+	}
+	
+	private static void inject(LootTableLoadEvent event, Set<ResourceLocation> lootTableSet, ResourceLocation injectionLootTable, String address)
+	{
+		if(lootTableSet.contains(event.getName()))
 		{
-			LootPool pool = LootPool.lootPool().add(LootTableReference.lootTableReference(MSLootTables.DUNGEON_LOOT_INJECT)).name("dungeon_loot_inject").build();
+			LootPool pool = LootPool.lootPool().add(LootTableReference.lootTableReference(injectionLootTable)).name(address).build();
 			event.getTable().addPool(pool);
 		}
 	}
+	
+	@SubscribeEvent
+	public static void addCustomVillagerTrade(VillagerTradesEvent event)
+	{
+		Int2ObjectMap<List<VillagerTrades.ItemListing>> trades = event.getTrades();
+		
+		if(event.getType() == VillagerProfession.CARTOGRAPHER)
+		{
+			trades.get(1).add((villager, random) -> new MerchantOffer(new ItemStack(Items.EMERALD, 8), new ItemStack(Items.COMPASS), createFrogTempleMap(villager), 12, 7, 0.2F));
+		}
+		
+		if(event.getType() == VillagerProfession.LEATHERWORKER)
+		{
+			trades.get(3).add((villager, random) -> new MerchantOffer(new ItemStack(Items.EMERALD, 5), new ItemStack(MSItems.CRUMPLY_HAT.get()), 5, 3, 0.2F));
+		}
+		
+		if(event.getType() == VillagerProfession.MASON)
+		{
+			trades.get(2).add((villager, random) -> new MerchantOffer(new ItemStack(Items.EMERALD, 3), new ItemStack(MSItems.CARVING_TOOL.get()), 4, 2, 0.05F));
+			trades.get(2).add((villager, random) -> new MerchantOffer(new ItemStack(Items.EMERALD, 2), new ItemStack(MSItems.STONE_TABLET.get()), 6, 2, 0.05F));
+		}
+		
+		if(event.getType() == VillagerProfession.WEAPONSMITH)
+		{
+			trades.get(2).add((villager, random) -> createEnchantedItemOffer(random, new ItemStack(MSItems.LUCERNE_HAMMER.get()), 4, 3, 1, 0.05F));
+		}
+		
+		if(event.getType() == VillagerProfession.LIBRARIAN)
+		{
+			ItemStack[] mediumBookTypes = new ItemStack[]{new ItemStack(MSItems.NONBINARY_CODE.get()), new ItemStack(MSItems.BINARY_CODE.get()), new ItemStack(MSItems.TILLDEATH_HANDBOOK.get()), new ItemStack(MSItems.TABLESTUCK_MANUAL.get()), new ItemStack(MSItems.WISEGUY.get()), new ItemStack(MSItems.SASSACRE_TEXT.get()), new ItemStack(MSItems.FLARP_MANUAL.get())};
+			ItemStack[] overworldBookTypes = new ItemStack[]{new ItemStack(MSItems.COMPLETED_SBURB_CODE.get())};
+			trades.get(2).add((villager, random) -> new MerchantOffer(MSDimensions.isInMedium(villager.getServer(), villager.level.dimension()) ? mediumBookTypes[random.nextInt(mediumBookTypes.length)] : overworldBookTypes[random.nextInt(overworldBookTypes.length)], new ItemStack(Items.EMERALD, 3), 4, 2, 0.05F));
+		}
```

Comment:
@kirderf1 do you have any recommendations on how to reduce the frequency of a given trade? The only method I can see to reduce the frequency is to set these as a trade for an index which has a lot of other options. However in the case of the cartographer, the rate at which frog temple maps are a trade option is over 50%. Im going to change it so it shares the same index with the ocean monument and woodland mansion to avoid the issue but if you have advice it may be useful for the future

Reply:
So from the code and minecraft wiki, it looks like trading works something like this:
Villagers go through 5 different levels, and get up to two trade offers for each level, with each level for each profession having different possible trade offers.
When choosing offers, if there are two or less possible offers, it chooses that one or two offers. Otherwise it chooses two offers randomly.
Cartographer level 2 has a glass pane trade and ocean explorer map, which would normally choose those two every time. With the frog temple map, it'll choose two among the three, giving each item a probability of 2/3 to appear.

I think there are three different approaches that can be taken:
- Accept that offers added by us may contend with few other offers and thus have a high probability of appearing (what offers would we add that actually should have a lower probability of appearing?)
- Try to add offers to levels which have a lot of offers already, such as cartographer 4 or mason 4.
- Add copies of the vanilla trade offers to increase the probability of those. While this is an option, I don't think it's a good option as it goes against the weight-free design of trade offers.

Reply:
moved it to the second level trade